sensor/FlowUploadSensor.js

```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    schedule() {
+        let endTime = new Date() / 1000 - this.config.offset
+        //upload flow to cloud
+        log.info("try to upload flows from "
+         + this.startTime + "(" + new Date(this.startTime * 1000).toUTCString() + ")" + 
+         " to " + endTime + "(" + new Date(endTime * 1000).toUTCString() + ")")
+
+         return async(() => {
+            try {
+                let flows = await(this.getAllFlows(this.startTime, endTime))
+                //set next start point
+                this.startTime = endTime + 0.001
+                if (flows != null && flows.flows != null && Object.keys(flows.flows).length > 0) {
+                    let data = JSON.stringify(flows)
+                    //log.info("original:" + data)
+                    log.info("original length:" + data.length)
+                    let compressedData = await(this.compressData(data))
+                    //log.info("compressed:" + compressedData)
+                    let length = compressedData.length
+                    log.info("compressed length:" + length)
+                    if (length > this.config.maxLength) {
+                        log.warn("data length " + length + " exceeded max length " + this.config.maxLength + ", abort uploading to cloud")
+                    } else {
+                        this.uploadData(compressedData)
+                    }
+                } else {
+                    log.info("empty flows, wait to next round")
+                }
+            } catch (err) {
+                log.error("upload flows failed:" + err.toString())
+                //skip this time range, set next start point
+                this.startTime = endTime + 0.001
+            }
+          })();
+    }
+
+    uploadData(data) {
+        let toUpload = {
+            payload : data
+        }    
+        Bone.flowgraph('flow', toUpload, function(err, response){
+            if (err) {
+                log.error("upload to cloud failed:" + err)
+            } else {
+                log.info("upload to cloud success:" + JSON.stringify(response))
+            }
+        })
+    }
+
+    compressData(data) {
+        return async(() => {
+            return new Promise(function (resolve, reject) {
+                let input = new Buffer(data, 'utf8');
+                zlib.deflate(input, (err, output) => {
+                    if(err) {
+                        reject(err)
+                    } else {
+                        resolve(output.toString('base64'))
+                    }
+                })
+            })
+        })();
+    }
+
+    getAllFlows(start, end) {
+        return async(() => {
+            let macs = await(this.getAllMacs())
+            let flows = {}
+            macs.forEach((mac) => {
+                let flow = await(this.getFlowByMac(mac, start, end))
+                if (flow != null && flow.length > 0) {
+                    flows[flowUtil.hashMac(mac)] = this.processFlow(flow, true)
```

Comment:
let hashdebug = sysManager.isSystemDebugOn();
        let _flow = flowUtil.hashFlow(flow,!hashdebug);

This way, if debug is enabled, or people using master, we get some clear text data.   We can do debug. Otherwsie, we can't read